### PR TITLE
FFS with FTP over LittleFS

### PR DESCRIPTION
**Pull request by jigglypuff-mchp**
Hi Hemanth,

- Integrated FTP over LittleFS file system (Network Drive project) to existing FFS demo for PIC32MZW1
- Added modified readme file with new console output and FTP server log snaps

Please review the changes and let me know your thoughts.

-Madhura

### CODE REVIEW

**Comment by htenkod** [comment]
> Hi Hemanth,
> 
> * Integrated FTP over LittleFS file system (Network Drive project) to existing FFS demo for PIC32MZW1
> * Added modified readme file with new console output and FTP server log snaps
> 
> Please review the changes and let me know your thoughts.
> 
> -Madhura

Hi Madhura,

Thanks for integrating and sharing the changes!

Few generic comments:- 

1. Make changes to auto detect the ECC with a basic API and use ECC storage as default mode
2. adduser and deleteuser are modifying the existing user credentials without authenticating the identity, so I suggest to use the file based user modification over the FTP link. It has 2 advantages, 1) Remote(wireless) modification of user details 2) Authentication of user
But if you move to file based user credential updates, in case of ECC, just after you take/read the info from userinfo file, clear (make username/password as ***** ) the info in the file so at least incase of ECC there is no way to fetch the details by tapping the external SPI flash.

3. FTP should be enabled only in the configuration mode, now it looks like even after connecting to home AP (After FFS) the device can be accessed with FTP

Regards,
Hemanth

File: Example/wifi_sta/firmware/src/app.c [hunk 1]
```diff
@@ -75,7 +114,176 @@ uint8_t CACHE_ALIGN work[SYS_FS_FAT_MAX_SS];
...
+static inline void _Command_setECC(SYS_CMD_DEVICE_NODE* pCmdIO, int argc, char** argv){
+    
+    setECC = true;                          //ECC will be used to store FTP server login credentials
+    appData.state = APP_ECC_Init;
+}
+
+static int _Command_AddUser(SYS_CMD_DEVICE_NODE* pCmdIO, int argc, char** argv){
+    
+    if(setECC){
+        clientLoginInfo client;
```
Comment (htenkod) [inline Example/wifi_sta/firmware/src/app.c]: This is a local variable and the buffer's inside can have a uninitialized values, now if you do strncpy it doesnt guarantee that the destination buffer gets a string termination char '\0' at the end of the string. the validateuserPass can find a string of length more than given length.

Example, if the user passes a 5 bytes string (username or password) then if the local variable has a non zero (which is very common) values after 5 bytes, the string length would be different than the passed 5 bytes.

File: Example/wifi_sta/firmware/src/app.c [hunk 2]
```diff
@@ -75,7 +114,176 @@ uint8_t CACHE_ALIGN work[SYS_FS_FAT_MAX_SS];
...
+        
+        strncpy(client.username, argv[1], 10);
+        strncpy(client.password, argv[2], 10);
+        if(validateUserPass(client.username, client.password))
```
Comment (htenkod) [inline Example/wifi_sta/firmware/src/app.c]: If the username length is 10 bytes, then validateUserPass strlen(user) will be added with the password length.
The username and password buffer are contagious memory and strncpy copies 10 bytes without the space for the null string character.

File: Example/wifi_sta/firmware/src/app.c [hunk 3]
```diff
@@ -75,7 +114,176 @@ uint8_t CACHE_ALIGN work[SYS_FS_FAT_MAX_SS];
...
+            char tempRead[32]; 
+            clientLoginInfo *tempRead2;
+            tempRead2 = (clientLoginInfo *)tempRead;
+            
+            status = atcab_read_zone(ECC_TNG_DATA_ZONE, ECC_TNG_GP_DATA_SLOT, ECC_TNG_GP_DATA_SLOT_BLOCK_BASE, 0, tempRead, 32);
+            if (ATCA_SUCCESS == status) 
+            {
+                if(0 == strncmp(client1.username, tempRead2->username, sizeof(tempRead2->username)))
+                {
+                    atcab_write_zone(ECC_TNG_DATA_ZONE, ECC_TNG_GP_DATA_SLOT, ECC_TNG_GP_DATA_SLOT_BLOCK_BASE, 0, (uint8_t*)resetDefault, 32);
```
Comment (htenkod) [inline Example/wifi_sta/firmware/src/app.c]: user default writes resetDefault string into the ECC which is nothing but having an user with user name 'FFFFFFFF' and the older password

The right way to make use of ECC space is to have a magic number (unique bytes) to determine if the details are valid or not. To make it invalid write a invalid magic number

File: Example/wifi_sta/firmware/src/app.c [hunk 4]
```diff
@@ -680,35 +925,49 @@ void APP_Tasks ( void )
...
         case APP_STATE_SERVICE_TASKS:
         {
-
+            
+            break;
+        } 
+        case APP_ECC_Init:
+        {
+            extern ATCAIfaceCfg atecc608_0_init_data;
+            if (ATCA_SUCCESS == atcab_init(&atecc608_0_init_data))
```
Comment (htenkod) [inline Example/wifi_sta/firmware/src/app.c]: Add initialization as part of APP init and enable the ecc method or non-ecc method

File: Example/wifi_sta/firmware/src/app.c [hunk 5]
```diff
@@ -38,25 +39,63 @@
...
+bool setECC = false;
 
-APP_DATA appData;
+// *****************************************************************************
+// *****************************************************************************
+// Section: Type Definitions
+// *****************************************************************************
+// *****************************************************************************
+/*struct used to store FTP server login credentials - username and password*/
+typedef struct client{
```
Comment (htenkod) [inline Example/wifi_sta/firmware/src/app.c]: Good to have structure definition in the header file

File: Example/wifi_sta/firmware/src/app.c [hunk 6]
```diff
@@ -75,7 +114,176 @@ uint8_t CACHE_ALIGN work[SYS_FS_FAT_MAX_SS];
...
+                {
+                    /*Parse the file */
+                    cJSON *messageJson = cJSON_Parse(fileData);
+                    if (messageJson == NULL) {
+                        const char *error_ptr = cJSON_GetErrorPtr();            
+                        if (error_ptr != NULL) {
+                            SYS_CONSOLE_PRINT("Message JSON parse Error. Error before: %s \r\n", error_ptr);
+                        }
+                        cJSON_Delete(messageJson);
+                        return false;
```
Comment (htenkod) [inline Example/wifi_sta/firmware/src/app.c:161]: Allocated memory is not freed

File: Example/wifi_sta/firmware/src/app.c [hunk 7]
```diff
@@ -75,7 +114,176 @@ uint8_t CACHE_ALIGN work[SYS_FS_FAT_MAX_SS];
...
+                    }
+                    
+                    cJSON *username = cJSON_GetObjectItem(messageJson, "Username");
+                    if (!username || username->type !=cJSON_String ) {
+                        SYS_CONSOLE_MESSAGE("Username parsing error\r\n");
+                        cJSON_Delete(messageJson);
+                        return false;
```
Comment (htenkod) [inline Example/wifi_sta/firmware/src/app.c:168]: Allocated memory is not freed

File: Example/wifi_sta/firmware/src/app.c [hunk 8]
```diff
@@ -75,7 +114,176 @@ uint8_t CACHE_ALIGN work[SYS_FS_FAT_MAX_SS];
...
+                    }
+                    
+                    cJSON *pw = cJSON_GetObjectItem(messageJson, "Password");
+                    if (!pw || pw->type !=cJSON_String ) {
+                        SYS_CONSOLE_MESSAGE("Password parsing error\r\n");
+                        cJSON_Delete(messageJson);
+                        return false;
```
Comment (htenkod) [inline Example/wifi_sta/firmware/src/app.c:175]: Allocated memory is not freed

**Review by htenkod — COMMENTED** [review]
Please go through the review comments and make the necessary changes

File: Example/wifi_sta/firmware/src/app.c (hunk 5 above)
Comment (jigglypuff-mchp) [reply Example/wifi_sta/firmware/src/app.c]: done

File: Example/wifi_sta/firmware/src/app.c (hunk 1 above)
Comment (jigglypuff-mchp) [reply Example/wifi_sta/firmware/src/app.c]: removed the function

File: Example/wifi_sta/firmware/src/app.c (hunk 2 above)
Comment (jigglypuff-mchp) [reply Example/wifi_sta/firmware/src/app.c]: removed the function

File: Example/wifi_sta/firmware/src/app.c (hunk 3 above)
Comment (jigglypuff-mchp) [reply Example/wifi_sta/firmware/src/app.c]: removed the function

File: Example/wifi_sta/firmware/src/app.c (hunk 6 above)
Comment (jigglypuff-mchp) [reply Example/wifi_sta/firmware/src/app.c:161]: done

File: Example/wifi_sta/firmware/src/app.c (hunk 7 above)
Comment (jigglypuff-mchp) [reply Example/wifi_sta/firmware/src/app.c:168]: done

File: Example/wifi_sta/firmware/src/app.c (hunk 4 above)
Comment (jigglypuff-mchp) [reply Example/wifi_sta/firmware/src/app.c]: done

**Comment by jigglypuff-mchp** [comment]
Hi Hemanth,
Please review the changes submitted on dev branch (3 latest commits). Thanks.

**Comment by jigglypuff-mchp** [comment]
reopening pull request for review

**Comment by htenkod** [comment]
The latest PR are based on earlier changes so I have to merge all the changes to master to avoid mismatches.
So, make the changes I requested and submit the PR through a different branch (based on master) and have 2 seperate commits, one for the MCC generated files (config folder) and other for the application changes.

**Pull request closed without merging**